crates/kornia-tracking/src/image_utilities.rs

```diff
@@ -0,0 +1,147 @@
+//! Image utilities used by the kornia-tracking modules.
+//! This module contains functions to compute image gradients, check bounds,
+//! perform the SE(2) exponential map, and detect keypoints.
+
+use image::{GenericImageView, GrayImage};
+use imageproc::corners::{corners_fast9, Corner};
+use glam::{Mat3, Vec3};
+
+/// Computes image gradients at (x, y).
+pub fn image_grad(grayscale_image: &GrayImage, x: f32, y: f32) -> Vec3 {
+    let ix = x.floor() as u32;
+    let iy = y.floor() as u32;
+    let dx = x - ix as f32;
+    let dy = y - iy as f32;
+    let ddx = 1.0 - dx;
+    let ddy = 1.0 - dy;
+
+    let px0y0 = grayscale_image.get_pixel(ix, iy).0[0] as f32;
+    let px1y0 = grayscale_image.get_pixel(ix + 1, iy).0[0] as f32;
+    let px0y1 = grayscale_image.get_pixel(ix, iy + 1).0[0] as f32;
+    let px1y1 = grayscale_image.get_pixel(ix + 1, iy + 1).0[0] as f32;
+    let res0 = ddx * ddy * px0y0 + ddx * dy * px0y1 + dx * ddy * px1y0 + dx * dy * px1y1;
+
+    let pxm1y0 = grayscale_image.get_pixel(ix - 1, iy).0[0] as f32;
+    let pxm1y1 = grayscale_image.get_pixel(ix - 1, iy + 1).0[0] as f32;
+    let res_mx = ddx * ddy * pxm1y0 + ddx * dy * pxm1y1 + dx * ddy * px0y0 + dx * dy * px0y1;
+
+    let px2y0 = grayscale_image.get_pixel(ix + 2, iy).0[0] as f32;
+    let px2y1 = grayscale_image.get_pixel(ix + 2, iy + 1).0[0] as f32;
+    let res_px = ddx * ddy * px1y0 + ddx * dy * px1y1 + dx * ddy * px2y0 + dx * dy * px2y1;
+    let res1 = 0.5 * (res_px - res_mx);
+
+    let px0ym1 = grayscale_image.get_pixel(ix, iy - 1).0[0] as f32;
+    let px1ym1 = grayscale_image.get_pixel(ix + 1, iy - 1).0[0] as f32;
+    let res_my = ddx * ddy * px0ym1 + ddx * dy * px0y0 + dx * ddy * px1ym1 + dx * dy * px1y0;
+
+    let px0y2 = grayscale_image.get_pixel(ix, iy + 2).0[0] as f32;
+    let px1y2 = grayscale_image.get_pixel(ix + 1, iy + 2).0[0] as f32;
+    let res_py = ddx * ddy * px0y1 + ddx * dy * px0y2 + dx * ddy * px1y1 + dx * dy * px1y2;
+    let res2 = 0.5 * (res_py - res_my);
+
+    Vec3::new(res0, res1, res2)
+}
+
+/// Checks if a keypoint is within the acceptable boundaries of the image.
+pub fn point_in_bound(keypoint: &Corner, height: u32, width: u32, radius: u32) -> bool {
+    keypoint.x >= radius
+        && keypoint.x <= width - radius
+        && keypoint.y >= radius
+        && keypoint.y <= height - radius
+}
+
+/// Checks if a coordinate (x, y) is within image bounds, considering a radius.
+pub fn inbound(image: &GrayImage, x: f32, y: f32, radius: u32) -> bool {
+    let x = x.round() as u32;
+    let y = y.round() as u32;
+    x >= radius && y >= radius && x < image.width() - radius && y < image.height() - radius
+}
+
+/// Computes the SE(2) exponential map based on a Vec3 input.
+pub fn se2_exp_matrix(a: &Vec3) -> Mat3 {
```

Comment:
this probably could be the starting point for a kornia-liealg crate (i'll open a separated ticket for this)
ref: https://kornia.readthedocs.io/en/latest/geometry.liegroup.html